Detectors/TRD/calibration/src/CalibratorGain.cxx

```diff
@@ -0,0 +1,206 @@
+// or submit itself to any jurisdiction.
+
+/// \file CalibratorGain.cxx
+/// \brief TimeSlot-based calibration of gain
+/// \author Ole Schmidt, Gauthier Legras
+
+#include "TRDCalibration/CalibratorGain.h"
+#include "Framework/ProcessingContext.h"
+#include "Framework/TimingInfo.h"
+#include "Framework/InputRecord.h"
+#include "DataFormatsTRD/Constants.h"
+#include "TRDBase/GeometryBase.h"
+#include "TStopwatch.h"
+#include "CCDB/CcdbApi.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "CommonUtils/NameConf.h"
+#include "CommonUtils/MemFileHelper.h"
+#include <TFile.h>
+#include <TTree.h>
+#include <TF1Convolution.h>
+
+#include <string>
+#include <map>
+#include <memory>
+#include <ctime>
+
+using namespace o2::trd::constants;
+
+namespace o2::trd
+{
+
+using Slot = o2::calibration::TimeSlot<GainCalibHistos>;
+
+void CalibratorGain::initOutput()
+{
+  // reset the CCDB output vectors
+  mInfoVector.clear();
+  mObjectVector.clear();
+}
+
+void CalibratorGain::initProcessing()
+{
+  if (mInitDone) {
+    return;
+  }
+  for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+    mdEdxhists[iDet] = std::make_unique<TH1F>(Form("hdEdx%d", iDet), "dEdx", NBINSGAINCALIB, 0., NBINSGAINCALIB);
+  }
+
+  // Init fitting functions
+  fConv = new TF1Convolution("[3] * TMath::Landau(x, [0], [1]) * exp(-[2]*x)", "TMath::Gaus(x, 0, [0])", 0., NBINSGAINCALIB, true);
+  fConv->SetNofPointsFFT(2000);
+  fitFunction = new TF1("fitConvLandau", *fConv, 0., NBINSGAINCALIB, fConv->GetNpar());
+  fitFunction->SetParameters(40, 15, 0.02, 1, 0.1);
+  fitFunction->SetParLimits(0, 30., 100.0);
+  fitFunction->SetParLimits(1, 5.0, 25.0);
+  fitFunction->SetParLimits(2, -0.1, 0.5);
+  fitFunction->SetParLimits(3, 1.0, 10.0);
+  fitFunction->SetParLimits(4, -0.1, 0.5);
+
+  // set tree addresses
+  if (mEnableOutput) {
+    mOutTree->Branch("MPVdEdx", &fitResults);
+    for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+      mOutTree->Branch(Form("dEdxHist_%d", iDet), mdEdxhists[iDet].get());
+    }
+  }
+
+  mInitDone = true;
+}
+
+void CalibratorGain::retrievePrev(o2::framework::ProcessingContext& pc)
+{
+
+  fitResults.fill(constants::MPVDEDXDEFAULT);
+  // We either get a pointer to a valid object from the last ~hour or to the default object
+  // which is always present. The first has precedence over the latter.
+  /*auto dataCalGain = pc.inputs().get<o2::trd::CalGain*>("calgain");
+  std::string msg = "Default Object";
+  // We check if the object we got is the default one by comparing it to the defaults.
+  for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+    if (dataCalGain->getMPVdEdx(iDet) != constants::MPVDEDXDEFAULT) {
+      msg = "Previous Object";
+      break;
+    }
+  }
+  LOG(info) << "Calibrator: From CCDB retrieved " << msg;
+
+  // Here we set each entry regardless if it is the default or not.
+  for (int iDet = 0; iDet < MAXCHAMBER; ++iDet) {
+    fitResults[iDet] = dataCalGain->getMPVdEdx(iDet);
+  }*/
```

Comment:
I was thinking that uploading the previous value when not enough statistics should be better that uploading the default value.
But in fact, we could also decide to do an interpolation, so maybe it is better to let it commented (or even remove it and let the default value) and discuss it later.

Reply:
I would not complicate things with an interpolation. Taking the value from the previous calibration is OK. As said, we can upload a default value beforehand and then you can uncomment this block already now

Reply:
Hi @martenole, in this case can you upload the object already now or tell me how to do it, so that I can uncomment and test the code before updating the PR?